Datastore/src/Query/AggregationQuery.php

```diff
@@ -0,0 +1,147 @@
+ * $query->filter('companyName', '=', 'Google');
+ * $query->addAggregation(Aggregation::count()->alias('total_upto_100'));
+ *
+ * $res = $datastore->runAggregationQuery($query);
+ * echo $res->get('total_upto_100');
+ * ```
+ *
+ * @see https://cloud.google.com/datastore/reference/rest/v1/projects/runAggregationQuery#query
+ * Aggregation Query Object Reference
+ * @see https://cloud.google.com/datastore/docs/aggregation-queries Aggregation Queries
+ */
+class AggregationQuery
+{
+    const TYPE_COUNT = 'count';
+
+    /**
+     * @var QueryInterface|null
+     */
+    private $query;
+
+    /**
+     * @var array
+     */
+    private $aggregates = [];
+
+    /**
+     * Create an aggregation query.
+     *
+     * @param QueryInterface|null $query
+     */
+    public function __construct($query = null)
+    {
+        $this->query = $query;
+    }
+
+    /**
+     * Adds a Query Aggregation.
+     *
+     * Accepts an array of properties for aggregation.
+     *
+     * Example:
+     * ```
+     *
+     * $query = $datastore->AggregationQuery();
+     * $query->kind('Companies');
+     * $query->filter('companyName', '=', 'Google');
+     * $query->addAggregation(Aggregation::count()->alias('total'));
+     * echo json_encode($query->queryObject());
+     * ```
+     *
+     * @param Aggregation|null $aggregation The Aggregation to be included.
+     *        Should be null for GqlQuery.
+     * @return AggregationQuery
+     */
+    public function addAggregation($aggregation)
+    {
+        $this->aggregates[] = $aggregation->getProps();
+        return $this;
+    }
+
+    /**
+     * Set the Query Projection.
+     *
+     * Accepts an array of properties. If set, only these properties will be
+     * returned.
+     *
+     * Example:
+     * ```
+     * $query = $datastore->query();
+     * $query->kind('Companies');
+     * $query->filter('companyName', '=', 'Google');
+     *
+     * $pipeline = $datastore->AggregationQuery()
+     *     ->over($query)
+     *     ->addAggregation(Aggregation::count()->alias('total'));
+     * ```
+     *
+     * @param QueryInterface $query The query whose properties to include.
+     * @return AggregationQuery
+     */
+    public function over($query)
+    {
+        $this->query = $query;
+        return $this;
+    }
+
+    public function queryObject()
+    {
+        if ($this->query instanceof Query) {
+            return [
+                'aggregationQuery' => [
+                    'nestedQuery' => $this->query->queryObject(),
+                    'aggregations' => $this->aggregates,
+                ]
+            ];
+        } elseif ($this->query instanceof GqlQuery) {
+            return [
+                'gqlQuery' => $this->query->queryObject(),
+            ];
+        } else {
+            throw new \UnexpectedValueException('unknown query type');
+        }
+    }
+
+    /**
+     * Return the query_type union field name.
+     *
+     * @return string
+     * @access private
```

Comment:
`queryKey()` method is only used internally in `Operation->runQuery(..)` to distinguish among various query types (Aggregated / Standard Query / GqlQuery) while forming the underlying request.
It's a pattern we've followed since the Query class was implemented, although not ideal but changing it will be messy.

Another approach I considered was `AggregationQuery implements QueryInterface`, but that forced to implement `canPaginate` which wasn't needed.

Reply:
I do not think this is being used. I looked through the code, and the only place `queryKey` is called is in `runQuery`, which requires `QueryInterface`. So this class is never passed to it (it's instead passed to `runAggregationQuery`). 

Furthermore, I removed this function all-together (along with the calls to mocking it in the tests), and all the tests pass. So I think we can remove this method!

Reply:
Removed, nice catch